# sequences and biopython
## 9/23/2021

<a href="?print-pdf">print view</a>

In [2]:
%%html

<script src="http://bits.csb.pitt.edu/asker.js/lib/asker.js"></script>

<script>

require(['https://cdnjs.cloudflare.com/ajax/libs/Chart.js/2.2.2/Chart.js'], function(Ch){
 Chart = Ch;
});

$('head').append('<link rel="stylesheet" href="http://bits.csb.pitt.edu/asker.js/themes/asker.default.css" />');


//the callback is provided a canvas object and data 
var chartmaker = function(canvas, labels, data) {
  var ctx = $(canvas).get(0).getContext("2d");
     var dataset = {labels: labels,                     
    datasets:[{
     data: data,
     backgroundColor: "rgba(150,64,150,0.5)",
         fillColor: "rgba(150,64,150,0.8)",    
  }]};
  var myBarChart = new Chart(ctx,{type:'bar',data:dataset,options:{legend: {display:false},scales:{yAxes:[{ticks:{min: 0}}]}}});


};

$(".jp-InputArea .o:contains(html)").closest('.jp-InputArea').hide();

</script>

# Sequence data

[http://www.ncbi.nlm.nih.gov](http://www.ncbi.nlm.nih.gov)

[Example](http://www.ncbi.nlm.nih.gov/gene/5216)

# FASTA

First line is description of sequence and starts with `>`

All lines up to the next `>` are part of the same sequence

Usually less than 80 characters per line

```
>gi|568815581:c4949086-4945650 Homo sapiens chromosome 17, GRCh38.p2 Primary Assembly
CCCGCAGGGTCCACACGGGTCGGGCCGGGCGCGCTCCCGTGCAGCCGGCTCCGGCCCCGACCGCCCCATG
CACTCCCGGCCCCGGCGCAGGCGCAGGCGCGGGCACACGCGCCGCCGCCCGCCGGTCCTTCCCTTCGGCG
GAGGTGGGGGAAGGAGGAGTCATCCCGTTTAACCCTGGGCTCCCCGAACTCTCCTTAATTTGCTAAATTT
GCAGCTTGCTAATTCCTCCTGCTTTCTCCTTCCTTCCTTCTTCTGGCTCACTCCCTGCCCCGATACCAAA
GTCTGGTTTATATTCAGTGCAAATTGGAGCAAACCCTACCCTTCACCTCTCTCCCGCCACCCCCCATCCT
TCTGCATTGCTTTCCATCGAACTCTGCAAATTTTGCAATAGGGGGAGGGATTTTTAAAATTGCATTTGCA
```

# Genbank

Annotated format.  Starts with `LOCUS` field.  Can have several other annotation (e.g. `KEYWORDS`, `SOURCE`, `REFERENCE`, `FEATURES`).

`ORIGIN` record indicates start of sequence

'`\\`' indicates the end of sequence

```
LOCUS       CAG28598                 140 aa            linear   PRI 16-OCT-2008
DEFINITION  PFN1, partial [Homo sapiens].
ACCESSION   CAG28598
VERSION     CAG28598.1  GI:47115277
DBSOURCE    embl accession CR407670.1
KEYWORDS    .
SOURCE      Homo sapiens (human)
  ORGANISM  Homo sapiens
            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;
            Mammalia; Eutheria; Euarchontoglires; Primates; Haplorrhini;
            Catarrhini; Hominidae; Homo.
ORIGIN      
        1 magwnayidn lmadgtcqda aivgykdsps vwaavpgktf vnitpaevgv lvgkdrssfy
       61 vngltlggqk csvirdsllq dgefsmdlrt kstggaptfn vtvtktdktl vllmgkegvh
      121 gglinkkcye mashlrrsqy
//
```

# Biopython

*Biopython features include parsers for various Bioinformatics file formats (BLAST, Clustalw, FASTA, Genbank,...), access to online services (NCBI, Expasy,...), interfaces to common and not-so-common programs (Clustalw, DSSP, MSMS...), a standard sequence class, various clustering modules, a KD tree data structure etc. and even documentation.*

Other modules that might be of interest:

 *   Pycogent: http://pycogent.org/
 *   bx-python: http://bitbucket.org/james_taylor/bx-python/wiki/Home
 *   DendroPy: http://packages.python.org/DendroPy/
 *   Pygr: http://code.google.com/p/pygr/
 *   bioservices: https://bioservices.readthedocs.io/en/master/
 
Biopython is **not** for performing sequencing itself (see: https://crc.pitt.edu/training/fall-2021-next-generation-sequencing-workshops).

# Sequence Objects

In [3]:
from Bio.Seq import Seq # the submodule structure of biopython is a little awkward

s = Seq("GATTACA")
s

Seq('GATTACA')

Sequences act a lot like strings, but have an associated *alphabet* and additional methods.

Methods shared with `str`: `count`, `endswith`, `find`, `lower`, `lstrip`, `rfind`, `split`, `startswith`, `strip`,`upper`

`Seq` methods:`back_transcribe`, `complement`, `reverse_complement`, `tomutable`, `tostring`, `transcribe`, `translate`, `ungap`

# Accessing Seq data

Sequences act like strings (indexed from 0)

In [8]:
s[0]

'G'

In [9]:
s[2:4] #returns sequence

Seq('TT')

In [10]:
s.lower()

Seq('gattaca')

In [11]:
s + s

Seq('GATTACAGATTACA')

# The Central Dogma

```
DNA coding strand (aka Crick strand, strand +1)	 
5’	ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG	3’
 	|||||||||||||||||||||||||||||||||||||||	 
3’	TACCGGTAACATTACCCGGCGACTTTCCCACGGGCTATC	5’
 	DNA template strand (aka Watson strand, strand −1)	 
                        |	 
                    Transcription	 
                        ↓	 
 
5’	AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG	3’
 	Single stranded messenger RNA	 
                        |	 
                   Translation	 
                        ↓	 
                  MAIVMGR*KGAR*
        amino acid sequence (w/stop codons)
 ```

In [12]:
dna = Seq('GATTACAGATTACAGATTACA')
dna.complement(),dna.reverse_complement()

(Seq('CTAATGTCTAATGTCTAATGT'), Seq('TGTAATCTGTAATCTGTAATC'))

# The Central Dogma

In [13]:
dna

Seq('GATTACAGATTACAGATTACA')

In [14]:
rna = dna.transcribe()
rna

Seq('GAUUACAGAUUACAGAUUACA')

In [15]:
protein = rna.translate()
protein

Seq('DYRLQIT')

In [16]:
dna.translate() #unlike cells, don't actually need rna

Seq('DYRLQIT')

# Codon Tables

By default the standard translation table is used, but others can be provided to the translate method.

In [17]:
from Bio.Data import CodonTable
print(sorted(CodonTable.unambiguous_dna_by_name.keys()))

['Alternative Flatworm Mitochondrial', 'Alternative Yeast Nuclear', 'Archaeal', 'Ascidian Mitochondrial', 'Bacterial', 'Balanophoraceae Plastid', 'Blastocrithidia Nuclear', 'Blepharisma Macronuclear', 'Candidate Division SR1', 'Cephalodiscidae Mitochondrial', 'Chlorophycean Mitochondrial', 'Ciliate Nuclear', 'Coelenterate Mitochondrial', 'Condylostoma Nuclear', 'Dasycladacean Nuclear', 'Echinoderm Mitochondrial', 'Euplotid Nuclear', 'Flatworm Mitochondrial', 'Gracilibacteria', 'Hexamita Nuclear', 'Invertebrate Mitochondrial', 'Karyorelict Nuclear', 'Mesodinium Nuclear', 'Mold Mitochondrial', 'Mycoplasma', 'Pachysolen tannophilus Nuclear', 'Peritrich Nuclear', 'Plant Plastid', 'Protozoan Mitochondrial', 'Pterobranchia Mitochondrial', 'SGC0', 'SGC1', 'SGC2', 'SGC3', 'SGC4', 'SGC5', 'SGC8', 'SGC9', 'Scenedesmus obliquus Mitochondrial', 'Spiroplasma', 'Standard', 'Thraustochytrium Mitochondrial', 'Trematode Mitochondrial', 'Vertebrate Mitochondrial', 'Yeast Mitochondrial']


In [18]:
print(CodonTable.unambiguous_dna_by_name['Standard'])

Table 1 Standard, SGC0

  |  T      |  C      |  A      |  G      |
--+---------+---------+---------+---------+--
T | TTT F   | TCT S   | TAT Y   | TGT C   | T
T | TTC F   | TCC S   | TAC Y   | TGC C   | C
T | TTA L   | TCA S   | TAA Stop| TGA Stop| A
T | TTG L(s)| TCG S   | TAG Stop| TGG W   | G
--+---------+---------+---------+---------+--
C | CTT L   | CCT P   | CAT H   | CGT R   | T
C | CTC L   | CCC P   | CAC H   | CGC R   | C
C | CTA L   | CCA P   | CAA Q   | CGA R   | A
C | CTG L(s)| CCG P   | CAG Q   | CGG R   | G
--+---------+---------+---------+---------+--
A | ATT I   | ACT T   | AAT N   | AGT S   | T
A | ATC I   | ACC T   | AAC N   | AGC S   | C
A | ATA I   | ACA T   | AAA K   | AGA R   | A
A | ATG M(s)| ACG T   | AAG K   | AGG R   | G
--+---------+---------+---------+---------+--
G | GTT V   | GCT A   | GAT D   | GGT G   | T
G | GTC V   | GCC A   | GAC D   | GGC G   | C
G | GTA V   | GCA A   | GAA E   | GGA G   | A
G | GTG V   | GCG A   | GAG E   | GGG G   | G
--+---------

<img src="../files/amino_acids.svg">

In [63]:
%%html
<div id="seqtrans" style="width: 500px"></div>
<script>
$('head').append('<link rel="stylesheet" href="http://bits.csb.pitt.edu/asker.js/themes/asker.default.css" />');

    var divid = '#seqtrans';
	jQuery(divid).asker({
	    id: divid,
	    question: "What amino acid does the codon <tt>CAT</tt> translate to?",
		answers: ['H',"Y","V","A","X"],
        extra: ['Histidine','Tyrosine','Valine','Alanine','Other'],
        server: "http://bits.csb.pitt.edu/asker.js/example/asker.cgi",
		charter: chartmaker})
    
$(".jp-InputArea .o:contains(html)").closest('.jp-InputArea').hide();


</script>

# SeqRecord

Sequence data is read/written as `SeqRecord` objects.

These objects store additional information about the sequence (name, id, description, features)

`SeqIO` reads sequence records: 
  * must specify format
  * `read` to read a file with a single record
  * `parse` to iterate over file with mulitple records

In [20]:
from Bio.SeqRecord import SeqRecord
from Bio import SeqIO
seq = SeqIO.read('../files/p53.gb','genbank')
seq

SeqRecord(seq=Seq('GATGGGATTGGGGTTTTCCCCTCCCATGTGCTCAAGACTGGCGCTAAAAGTTTT...GTG'), id='NG_017013.2', name='NG_017013', description='Homo sapiens tumor protein p53 (TP53), RefSeqGene (LRG_321) on chromosome 17', dbxrefs=[])

In [21]:
seqs = []
# https://asinansaglam.github.io/python_bio_2022/files/hydra.fasta
for s in SeqIO.parse('../files/hydra.fasta','fasta'):
    seqs.append(s)

In [22]:
len(seqs)

40

# Fetching sequences from the Internet

Biopython provides and interface to the [NCBI "Entrez" search engine](http://www.ncbi.nlm.nih.gov/sites/gquery)

The results of internet queries are returned as file-like objects.

In [23]:
from Bio import Entrez
Entrez.email = "dkoes@pitt.edu" #biopython forces you to provide your email
res = Entrez.read(Entrez.einfo()) # the names of all available databases
res

{'DbList': ['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'homologene', 'medgen', 'mesh', 'ncbisearch', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'biosystems', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']}

In [24]:
print(sorted(res['DbList']))

['annotinfo', 'assembly', 'biocollections', 'bioproject', 'biosample', 'biosystems', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'dbvar', 'gap', 'gapplus', 'gds', 'gene', 'genome', 'geoprofiles', 'grasp', 'gtr', 'homologene', 'ipg', 'medgen', 'mesh', 'ncbisearch', 'nlmcatalog', 'nuccore', 'nucleotide', 'omim', 'orgtrack', 'pcassay', 'pccompound', 'pcsubstance', 'pmc', 'popset', 'protein', 'proteinclusters', 'protfam', 'pubmed', 'seqannot', 'snp', 'sra', 'structure', 'taxonomy']


# ESearch

You can search any database for a given term and it will return the ids of all the relevant records.

In [25]:
result = Entrez.esearch(db='nucleotide', term='tp53')  #the result is a file-like object of the raw xml data
records = Entrez.read(result) #put into a more palatable form (dictionary)
print(records)

{'Count': '12751', 'RetMax': '20', 'RetStart': '0', 'IdList': ['2095557441', '2095547223', '2095541784', '2095533838', '2088687916', '595763516', '595763184', '56790941', '1890342052', '1677539426', '1677537377', '1519316331', '1388003379', '1169640623', '383209646', '187960039', '187960038', '167736374', '167736372', '154146208'], 'TranslationSet': [], 'TranslationStack': [{'Term': 'tp53[All Fields]', 'Field': 'All Fields', 'Count': '12751', 'Explode': 'N'}, 'GROUP'], 'QueryTranslation': 'tp53[All Fields]'}


There were <strike>2923</strike> <strike>7616</strike> <strike>8238</strike> <strike>10725</strike> <strike>11850</strike> 12751 hits, but by default only 20 are returned.  We can change this ([and other parameters](http://www.ncbi.nlm.nih.gov/books/NBK25499/#chapter4.ESearch)) by changing our search terms.

In [26]:
records = Entrez.read(Entrez.esearch(db='nucleotide', term='tp53', retmax=50))
records

{'Count': '12751', 'RetMax': '50', 'RetStart': '0', 'IdList': ['2095557441', '2095547223', '2095541784', '2095533838', '2088687916', '595763516', '595763184', '56790941', '1890342052', '1677539426', '1677537377', '1519316331', '1388003379', '1169640623', '383209646', '187960039', '187960038', '167736374', '167736372', '154146208', '148747364', '67078437', '2092262714', '2092262693', '2092262690', '2092222507', '2092222504', '2092222501', '2092220105', '2092218640', '2092203059', '2092187984', '2092184066', '2092184064', '2092176789', '2092176786', '2092176784', '2092176781', '2092173253', '2092157841', '2092144944', '2092144942', '2092144564', '2092133000', '2092132653', '2092132649', '2092116963', '2092101586', '2092080167', '2092080165'], 'TranslationSet': [], 'TranslationStack': [{'Term': 'tp53[All Fields]', 'Field': 'All Fields', 'Count': '12751', 'Explode': 'N'}, 'GROUP'], 'QueryTranslation': 'tp53[All Fields]'}

# EFetch

To get the full record for a given id, use `efetch`.  

Must provide `rettype`  ([available options](http://www.ncbi.nlm.nih.gov/books/NBK25499/table/chapter4.T._valid_values_of__retmode_and/?report=objectonly) include fasta and gb)

`retmode` can be text or xml.

In [27]:
#fetch the genbank file for the first id from our search
result = Entrez.efetch(db="nucleotide",id=records['IdList'][1],rettype="gb",retmode='text')
#parse into a seqrecord
p53 = SeqIO.read(result,'gb')

In [28]:
result

<_io.TextIOWrapper encoding='UTF-8'>

In [29]:
p53

SeqRecord(seq=Seq('ACGTGAGCCTGTCTCCTCAGCGTGCACACACTCACCTGCACACACCCTGTGAAC...CCT'), id='JAATJU010022300.1', name='JAATJU010022300', description='Microtus ochrogaster isolate LTLLF scaffold31, whole genome shotgun sequence', dbxrefs=['BioProject:PRJNA541330', 'BioSample:SAMN11605340'])

# Features

Genbank files are typically annotated with features, which refer to portions of the full sequence

`SeqRecord` objects track these features and you can extract the corresponding subsequence

*CDS* - coding sequence

In [30]:
p53.features

[SeqFeature(FeatureLocation(ExactPosition(0), ExactPosition(10587605), strand=1), type='source'),
 SeqFeature(FeatureLocation(ExactPosition(7559), ExactPosition(62343), strand=-1), type='gene'),
 SeqFeature(CompoundLocation([FeatureLocation(ExactPosition(62289), ExactPosition(62343), strand=-1), FeatureLocation(ExactPosition(46544), ExactPosition(46649), strand=-1), FeatureLocation(ExactPosition(43511), ExactPosition(43673), strand=-1), FeatureLocation(ExactPosition(33520), ExactPosition(33649), strand=-1), FeatureLocation(ExactPosition(31983), ExactPosition(32169), strand=-1), FeatureLocation(ExactPosition(30375), ExactPosition(30644), strand=-1), FeatureLocation(ExactPosition(28385), ExactPosition(28528), strand=-1), FeatureLocation(ExactPosition(25029), ExactPosition(25275), strand=-1), FeatureLocation(ExactPosition(22983), ExactPosition(23199), strand=-1), FeatureLocation(ExactPosition(22343), ExactPosition(22474), strand=-1), FeatureLocation(ExactPosition(20333), ExactPosition(204

# Extracting subsequences

In [31]:
cdsfeature = p53.features[3]
print(cdsfeature)

type: CDS
location: join{[62289:62343](-), [46544:46649](-), [43511:43673](-), [33520:33649](-), [31983:32169](-), [30375:30644](-), [28385:28528](-), [25029:25275](-), [22983:23199](-), [22343:22474](-), [20333:20411](-), [7559:7763](-)}
qualifiers:
    Key: codon_start, Value: ['1']
    Key: inference, Value: ['COORDINATES: Similar to AA sequence:UniProtKB:P59862']
    Key: locus_tag, Value: ['LTLLF_151105']
    Key: note, Value: ['Derived by automated computational analysis using gene prediction method: Glean and BGI proprietary modifications; BGI Gene ID: Vole_GLEAN_10028642; Expectation value: 0.0']
    Key: product, Value: ['Cadherin-like protein 26']
    Key: protein_id, Value: ['KAH0511053.1']
    Key: translation, Value: ['MAIRGCTWLLLLLLPLLQGQNHQPLRRSKRRWVLTTLELEEEDPGPFPKLVGELFNNMSNNVSLMYLLRGPGVDEFPEIGLFSIEDHQSGKVYVHRAVDREVTPSFLTAPWFTLIVRATDCGEPSLSSTATIHISVEEGNNHMPTFVEDHYEIQISEGEVAWGVLDLPVQDGDSPFTPAWRAQFHILDGNEEEHFDIMTDPETNRAVLNVIKPLDYETQAAHSLSIVVENQEPLFSCEEGQAQPSTKAMASTTVSVQV

The subsequence the feature refers to can be extracted from the original full sequence using the feature.

In [32]:
coding = cdsfeature.extract(p53) #pass the full record (p53) to the feature
coding

SeqRecord(seq=Seq('ATGGCCATAAGAGGCTGCACCTGGTTGCTGCTACTGCTGCTGCCGCTCCTGCAG...TAA'), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])

In [64]:
%%html
<div id="qualifiers" style="width: 500px"></div>
<script>
$('head').append('<link rel="stylesheet" href="http://bits.csb.pitt.edu/asker.js/themes/asker.default.css" />');

    var divid = '#qualifiers';
	jQuery(divid).asker({
	    id: divid,
	    question: "What is the type of cdsfeature.qualifiers?",
		answers: ['str',"int","set","list","dict"],
        server: "http://bits.csb.pitt.edu/asker.js/example/asker.cgi",
		charter: chartmaker})
    
$(".jp-InputArea .o:contains(html)").closest('.jp-InputArea').hide();


</script>

# BLAST!

Biopython let's you use [NCBI's BLAST](http://blast.ncbi.nlm.nih.gov/Blast.cgi) to search for similar sequences with `qblast` which has three required arguments:
 * **program**: blastn, blastp, blastx, tblastn, tblastx
 * **database**: see website
 * **sequence**: a sequence object

In [35]:
from Bio.Blast import NCBIWWW
result = NCBIWWW.qblast("blastn","nt",coding.seq,hitlist_size=250)
#result is a file-like object with xml in it - it can take a while to get results

In [36]:
from Bio.Blast import NCBIXML #for parsing xmls
blast_records = NCBIXML.read(result)

In [37]:
print(len(blast_records.alignments),len(blast_records.descriptions))

250 250


In [38]:
alignment = blast_records.alignments[0]
print(len(alignment.hsps))

3


In [40]:
hsp = alignment.hsps[0] #high scoring segment pairs
print('****Alignment****')
print('sequence:', alignment.title)
print('length:', alignment.length)
print('e value:', hsp.expect)
print(hsp.query[0:75] + '...')  # what we searched with
print(hsp.match[0:75] + '...')
print(hsp.sbjct[0:75] + '...')  # what we matched to

****Alignment****
sequence: gi|1490716901|ref|XM_026787265.1| PREDICTED: Microtus ochrogaster cadherin 26 (Cdh26), mRNA
length: 3083
e value: 0.0
ACTGCTCCTTGGTTCACTCTGATAGTCAGAGCGACGGATTGTGGAGAGCCATCCCTGTCATCCACAGCCACCATT...
|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||...
ACTGCTCCTTGGTTCACTCTGATAGTCAGAGCGACGGATTGTGGAGAGCCATCCCTGTCATCCACAGCCACCATT...


In [41]:
alignment = blast_records.alignments[-1] # get last alignment
hsp = alignment.hsps[0]
print('****Alignment****')
print('sequence:', alignment.title)
print('length:', alignment.length)
print('e value:', hsp.expect)
print(hsp.query[0:75] + '...')  # what we searched with
print(hsp.match[0:75] + '...')
print(hsp.sbjct[0:75] + '...')  # what we matched to


****Alignment****
sequence: gi|1909942461|dbj|AP023480.1| Homo sapiens DNA, chromosome 20, nearly complete genome
length: 62398354
e value: 1.1612e-43
TCAAGCCTTTGGATTATGAAACTCAAGCAGCCCACAGCCTCAGCATTGTTGTGGAGAATCAGGAGCCGCTCTTCT...
|| |||||||||||||||| ||||   |||| || ||||||| |||||| ||||||||| |||||  |||| |||...
TCCAGCCTTTGGATTATGAGACTCGCCCAGCGCAAAGCCTCATCATTGTCGTGGAGAATGAGGAGAGGCTCGTCT...


# Creating alignments

Biopython uses `clustal` to perform multiple alignments

The interface is very simple - you construct the `clustal` commandline and must read the result files

In [51]:
from Bio.Align.Applications import ClustalwCommandline
cline = ClustalwCommandline('clustalw', infile='../files/hydra.fasta',outfile='alignment.aln')

In [52]:
print(cline)

clustalw -infile=../files/hydra.fasta -outfile=alignment.aln


In [54]:
output = cline() #this calls the above, may take a while

In [55]:
!head alignment.aln

CLUSTAL 2.1 multiple sequence alignment


gi|164609123|gb|ABY62783.1|/1-      ------------MLNIAILCLSCYINYVSCLSLTSPAIIEEVVGRSVTIT
model1_F.pdb                        ------------------------------LSLTSPAIIEEVVGRSVTIT
gi|225423262|gb|ACN91137.1|/1-      ------------MLNIAILSLSCYINYVSCLSLTSPAIIEEVVGRSVTIT
gi|302171750|gb|ADK97776.1|/1-      ------------MLNIAILSLSCYINYVSCLSLTSPAIIEEVVGRSVTIT
gi|302171774|gb|ADK97788.1|/1-      ------------MLNIAILCLSCYINYVSCLSLTSSAIIEEVVNRSVTIT
gi|225423244|gb|ACN91128.1|/1-      ------------MLNIAILCLSCYINYVSCLSLTSSAIIEEVVNRSVTIT
gi|225423272|gb|ACN91142.1|/1-      ------------MLNIAILLLSCYINYVSSLSLTSPAIIEEVVNRSVTIT


# Alignments

`AlignIO` is used to read alignment files (must provide format)

In [56]:
from Bio import AlignIO
align = AlignIO.read('alignment.aln','clustal')

In [57]:
align

<<class 'Bio.Align.MultipleSeqAlignment'> instance (40 records of length 691) at 11ce929a0>

In [58]:
print(align)

Alignment with 40 rows and 691 columns
------------MLNIAILCLSCYINYVSCLSLTSPAIIEEVVG...QKK gi|164609123|gb|ABY62783.1|/1-
------------------------------LSLTSPAIIEEVVG...--- model1_F.pdb
------------MLNIAILSLSCYINYVSCLSLTSPAIIEEVVG...QKK gi|225423262|gb|ACN91137.1|/1-
------------MLNIAILSLSCYINYVSCLSLTSPAIIEEVVG...QKK gi|302171750|gb|ADK97776.1|/1-
------------MLNIAILCLSCYINYVSCLSLTSSAIIEEVVN...QKK gi|302171774|gb|ADK97788.1|/1-
------------MLNIAILCLSCYINYVSCLSLTSSAIIEEVVN...QKK gi|225423244|gb|ACN91128.1|/1-
------------MLNIAILLLSCYINYVSSLSLTSPAIIEEVVN...QKK gi|225423272|gb|ACN91142.1|/1-
------------MLNIAILCLSYYVNYVSCLSLTSPAIIEEVVG...QKK gi|302171742|gb|ADK97772.1|/1-
------------MLNIAILCLSYYVNYVSCLSLTSPAIIEEVVG...QKK gi|225423264|gb|ACN91138.1|/1-
------------MLNIAILSLSCYINYVSCLSLTSPAIIEEVVG...QKK gi|302171780|gb|ADK97791.1|/1-
------------MLNIAILCLSCYINYVSCVSLTSPAIIEEVVG...QKK gi|302171756|gb|ADK97779.1|/1-
------------MLNVVILCLSCYINYVSCLSLTSPAIIEEVVG...QKK gi|225423288|gb|ACN91150.1

# Slicing Alignments

Alignments are sliced just like `numpy` arrays

In [59]:
align[0] #first row

SeqRecord(seq=Seq('------------MLNIAILCLSCYINYVSCLSLTSPAIIEEVVGRSVTITYVTD...QKK'), id='gi|164609123|gb|ABY62783.1|/1-', name='<unknown name>', description='gi|164609123|gb|ABY62783.1|/1-', dbxrefs=[])

In [60]:
align[:,0] #first column

'------------M-----------------M---------'

In [61]:
print(align[:,0:3] )

Alignment with 40 rows and 3 columns
--- gi|164609123|gb|ABY62783.1|/1-
--- model1_F.pdb
--- gi|225423262|gb|ACN91137.1|/1-
--- gi|302171750|gb|ADK97776.1|/1-
--- gi|302171774|gb|ADK97788.1|/1-
--- gi|225423244|gb|ACN91128.1|/1-
--- gi|225423272|gb|ACN91142.1|/1-
--- gi|302171742|gb|ADK97772.1|/1-
--- gi|225423264|gb|ACN91138.1|/1-
--- gi|302171780|gb|ADK97791.1|/1-
--- gi|302171756|gb|ADK97779.1|/1-
--- gi|225423288|gb|ACN91150.1|/1-
MKL gi|225423270|gb|ACN91141.1|/1-
--- gi|302171766|gb|ADK97784.1|/1-
--- gi|225423256|gb|ACN91134.1|/1-
--- gi|302171748|gb|ADK97775.1|/1-
--- gi|302171772|gb|ADK97787.1|/1-
--- gi|225423294|gb|ACN91153.1|/1-
...
--- gi|302171764|gb|ADK97783.1|/1-


# And now for a brief foray into marine microbiology...

https://asinansaglam.github.io/python_bio_2022/files/hydra.fasta

<img src="imgs/sequence_project/sequence_project.001.png" width="100%">

<img src="imgs/sequence_project/sequence_project.002.png" width="100%">

<img src="imgs/sequence_project/sequence_project.003.png" width="100%">

<img src="imgs/sequence_project/sequence_project.004.png" width="100%">

<img src="imgs/sequence_project/sequence_project.005.png"  width="100%">

<img src="imgs/sequence_project/sequence_project.006.png"  width="100%">

<img src="imgs/sequence_project/sequence_project.007.png" width="100%">

<img src="imgs/sequence_project/sequence_project.008.png" width="100%">

# Project

We have a gene (alr2), but what part of the gene is responsible for allorecognition?

Given 179 sequences, how might we find out?

https://asinansaglam.github.io/python_bio_2022/files/hydra179.aln

Find the part of the sequence that changes the most.
 * Count number of distinct residues at each position. Plot.
 * Count number of distinct *subsequences* of length 10 at each position. Plot.

<img src="imgs/sequence_project/sequence_project.011.png">

In [ ]:
import matplotlib.pyplot as plt

def countatpos(seqs,pos):
        vals = set()
        for s in seqs:
                vals.add(s[pos:pos+10])
        return len(vals)

f = open("alr_align");

seqs = f.readlines()

xaxis = list()
yaxis = list()
for i in xrange(len(seqs[0])):
        xaxis.append(i)
        yaxis.append(countatpos(seqs,i))
print(zip(xaxis,yaxis))
plt.bar(xaxis,yaxis)
plt.show()